File: 2016/18.py
from utils import readSingleLine

line = readSingleLine("input_18.txt")

part = 1
if part == 1:
    steps = 39
else:
    steps = 399999
    
c = line.count(".")
for s in range(steps):
    newline = ""
    for i in range(len(line)):
        if i == 0:
            if (line[i] == "^" and line[i+1] == "^") or (line[i+1] == "^"):
                newline += "^"
            else:
                newline += "."
        elif i == len(line)-1:
            if (line[i] == "^" and line[i-1] == "^") or (line[i-1] == "^"):
                newline += "^"
            else:
                newline += "."
        else:
            if (line[i] == "^" and line[i-1] == "^" and line[i+1] == ".") or \
               (line[i] == "^" and line[i-1] == "." and line[i+1] == "^") or \
               (line[i-1] == "^" and line[i] == "." and line[i+1] == ".") or \
               (line[i+1] == "^" and line[i] == "." and line[i-1] == "."):
                newline += "^"
            else:
                newline += "."
    c += newline.count(".")
    line = newline

if part == 1:
    print ("🎄Part 1: {}".format(c))
else:
    print ("🎁🎄Part 2: {}".format(c))

File: 2016/18.jl
include("utils.jl")
using .AoCUtils

function safe(line, steps)
    c = count(c->c=='.', line)
    for s in 1:steps
        newline = ""
        for i in eachindex(line)
            if i == 1
                if (line[i] == '^' && line[i+1] == '^') || (line[i+1] == '^')
                    newline *= '^'
                else
                    newline *= '.'
                end
            elseif i == length(line)
                if (line[i] == '^' && line[i-1] == '^') || (line[i-1] == '^')
                    newline *= '^'
                else
                    newline *= '.'
                end
            else
                if (line[i] == '^' && line[i-1] == '^' && line[i+1] == '.') || 
                (line[i] == '^' && line[i-1] == '.' && line[i+1] == '^') ||
                (line[i-1] == '^' && line[i] == '.' && line[i+1] == '.') ||
                (line[i+1] == '^' && line[i] == '.' && line[i-1] == '.')
                    newline *= '^'
                else
                    newline *= '.'
                end
            end
        end
        c += count(c->c=='.', newline)
        line = newline
    end 
    println("🎁🎄 $c")
end

lines = readLines("input_18.txt")
safe(lines[1], 39)
safe(lines[1], 399999)

